packages/kuma-gui/src/app/data-planes/views/DataPlaneDetailView.vue

```diff
@@ -177,34 +183,35 @@
                   />
                   Inbounds
                 </template>
-                <ConnectionGroup
-                  type="inbound"
-                  data-testid="dataplane-inbounds"
+
+                <!-- if we are a builtin gateway proxy i.e. a 'gateway' proxy -->
+                <!-- use its first and only inbounds as a template  -->
+                <!-- and fill the inbounds out from the stats once they are loaded -->
+                <template
+                  v-for="inbounds in [props.data.dataplane.networking.type === 'gateway' ? Object.entries<any>(traffic?.inbounds ?? {}).reduce<DataplaneInbound[]>((prev, [key, value]) => {
+                    // As we are just 'finding' inbounds from stats without knowing them
+                    // from the dataplane overview API request, we will wrongly 'find'
+                    // the envoy admin inbound that is used for kumas /stats API. Ignore
+                    // the envoy admin inbound when we find it
+                    const port = key.split('_').at(-1)
+                    if(port === (props.data.dataplane.networking.admin?.port ?? '9901')) {
+                      return prev
+                    }
+                    return prev.concat([
+                      {
+                        ...props.data.dataplane.networking.inbounds[0],
+                        name: key,
+                        port: Number(port),
+                        protocol: ['http', 'tcp'].find(item => typeof value[item] !== 'undefined') ?? 'tcp',
+                        addressPort: `${props.data.dataplane.networking.inbounds[0].address}:${port}`,
+                      },
+                    ])
+                  }, []) : props.data.dataplane.networking.inbounds]"
+                  :key="inbounds"
```

Comment:
Would it make sense to move this out of the template? It's hard to read because of ternary and nested logic. I know that this just pops up due to some surrounding changes and this haven't been touched, but I still want to mention it and see what you think about it 🙂

Reply:
> Would it make sense to move this out of the template? It's hard to read because of ternary and nested logic.

I would say no 😅 

Ummm, I would encourage you to read https://copyconstruct.medium.com/small-functions-considered-harmful-91035d316c29. Whilst you know I don't like it when people quote things from "somebody somewhere on the internet", but this person I happen to highly respect, I think a lot of folks do.

Cindy's intro and quite a useful TLDR;

> Some people seem so enamored with small functions that the idea of abstracting any and every piece of logic that might seem even nominally complex into a separate function is something that is passionately advocated for.

> I’ve worked on codebases inherited from folks who’d internalized this idea to such an unholy extent that the end result was pretty hellish and entirely antithetical to all the good intentions the road to it was paved with. 

I think the sections headed "Loss of Locality" and "Hard to Read" are particularly relevant here. Both specifically argue against moving things out of the place where its being executed (more or less) in the name of making things less hard to read.

I have also worked on codebases inherited from folks who’d internalized this idea to such an unholy extent that the end result was pretty hellish and entirely antithetical to all the good intentions the road to it was paved with. It has taken us a long while and a lot of extremely hard work (not just typing code into a keyboard) to get to the more solid foundations of where we are today.

Just to give you a heads up, I would guess that unless there is something technically incorrect here (let me know if there is!), I'm inclined not to change anything especially within the scope of this work (as you've pointed out), but let me know what you think of the article, I think thats the most important point here 👍

Reply:
Thanks for sharing this article 🙂 I really like it and I fully agree with it and your position. I also think that too many abstractions and outsourcing is counter productive and actually makes it harder to understand the code. Especially if I have to open a lot of different files for the full picture.

> Just to give you a heads up, I would guess that unless there is something technically incorrect here (let me know if there is!), I'm inclined not to change anything especially within the scope of this work (as you've pointed out), but let me know what you think of the article, I think thats the most important point here 👍

There is nothing wrong with this on the technical perspective, it just popped up due to the surrounding changes and I don't expect any changes to this within the scope of this PR.

However, my point was not to create abstractions for this, I just meant to ask if there is a better location for this other than the template. If I understand it correctly, we are partially shaping some nested data here, which I think should be part of the data layer. Please correct me if I'm wrong 🙂 

---

Feel free to resolve this discussion any time. This is not meant to block the progress of the PR.

Reply:
> Thanks for sharing this article 🙂 I really like it and I fully agree with it and your position. I also think that too many abstractions and outsourcing is counter productive and actually makes it harder to understand the code. Especially if I have to open a lot of different files for the full picture.

🙌  awesome, glad you like the article.

> There is nothing wrong with this on the technical perspective, it just popped up due to the surrounding changes and I don't expect any changes to this within the scope of this PR.

Cool yeah, sorry I totally think we are on the same page here by the way (not technically problematic and anything else would be out of scope), it's not my favourite piece of code but there are reasons that it is like it is. Its a bit difficult to explain the reasons why it is like it is, and its probably a longer story than this but 1 sec lemme try:

---

> If I understand it correctly, we are partially shaping some nested data here, which I think should be part of the data layer.

This is almost 100% correct! But! (yes, you guessed it there's always a but! 😆 ). We are merging two pieces of data here and this fact is very important.

`DataSource` is designed to be composable, and its also designed to be used declaratively inside templates so that complex loading/erroring displays are also composable. In order to make displayable composable data sourcing/loading/erroring the best place to do that is in the template because its all about the visuals. I know you know quite a bit about how DataSource works so I'm pretty sure you follow the approach here.

In this particular instance, we really wanted to have the two HTTP calls we use here be **independently** loaded/error'ed, so the option of using say `Promise.all([firstCall, secondCall]).then(doSomeMerging)` from within the `sources.ts` file, and therefore then use our data layer to do data merging shaping is not ideal. If we did this the loading/erroring in the template would have to be one single thing, each call (with its loading/erroring) wouldn't be an "independent display".

Hopefully that last paragraph ^ makes sense. Let me know if not because its really important and what everything else hangs on that.

If I've explained the above well enough, you'll see that the merging is best done in the the template. Granted, we could make a function and move the code elsewhere and just have the call site in the template. But we are risking some of the points pointed out in the above article.

Lastly, all of this said, I'm a firm believer in making sure we understand what's required from a piece of code before we start to try to optimize it or make it reusable (or even move it elsewhere if that makes sense, despite the above article).

The piece of work here (so the old dataplane work, not the current scope of zones) had very many unknowns all the way from start to finish. I specifically wanted to treat it like prototyping work until we knew what we wanted. I think only now I'm starting to get a feel for what I think we want because I'm doing similar functionality elsewhere (zones). I'm so glad I didn't try to prematurely optimize or worry too much about small pieces of code like this because I've found out over the past few days that zones is very different in subtle ways and if I'd tried to organize this a little better already it most definitely would have been premature.

Anyway, thanks for asking the questions, its really nice to have the reason and opportunity to try and explain these things in PRs. Not only is it good to share thinking, and validate or (or sometimes not 😅 ) reasoning, but it acts as good documentation for future us 🙌

Reply:
Ohhh, now this makes sense to me 😄 I've completely missed the `DataSource` and just saw the `props.data`, so I thought there is just one data set 🙈 
I might be wrong, but with the declarative data fetching approach (which I like), I think there is currently no good way of merging two data sets from different sources in the data layer (or outside of the template) 🤔 

> Lastly, all of this said, I'm a firm believer in making sure we understand what's required from a piece of code before we start to try to optimize it or make it reusable (or even move it elsewhere if that makes sense, despite the above article).
>
> The piece of work here (so the old dataplane work, not the current scope of zones) had very many unknowns all the way from start to finish. I specifically wanted to treat it like prototyping work until we knew what we wanted. I think only now I'm starting to get a feel for what I think we want because I'm doing similar functionality elsewhere (zones). I'm so glad I didn't try to prematurely optimize or worry too much about small pieces of code like this because I've found out over the past few days that zones is very different in subtle ways and if I'd tried to organize this a little better already it most definitely would have been premature.
>
> Anyway, thanks for asking the questions, its really nice to have the reason and opportunity to try and explain these things in PRs. Not only is it good to share thinking, and validate or (or sometimes not 😅 ) reasoning, but it acts as good documentation for future us 🙌

Yes, fully agree 💯 Thanks for the patience and answering my questions 🙂

Reply:
>I might be wrong, but with the declarative data fetching approach (which I like), I think there is currently no good way of merging two data sets from different sources in the data layer (or outside of the template) 🤔

Well there is if you don't want to have independent loading/error states. You can use `Promise.all` then merge from within `sources.ts`, so thats possible if you need to do this in the future.

There are also other approaches but what we do here is by far the simplest and keeps the independent loading/errors states in a place where you expect to see them ("expect to see them" both from a user perspective and a code perspective)